### ISSUE TITLE
fromXML builtin?

### ISSUE THREAD

**Issue by amazari** [NixOS/nixpkgs#57239]
Hello fellow nixers,

While looking into building a Kodi-repository-ton-nix-packages-set converter, I have to parse an XML file containing the metadata of the repository. 
Ideally I'd like to parse it to a Nix value and apply transformations to turn it into my desired form.

AFAIK, the is no XML parsing facilities in the builtins. XML being still quite widespread, this is quite surprising. Surely other packages have the same requirements. Are they each  implementing an ad-hoc parser? Did I look in the wrong place?

The parser wouldn't need to be fully sax or dom conformant but be able to turn simple XML to Nix expression:

```
nixExpr = fromXML(''
<outer>
  <tag1>42</tag1>
  <tag2>
    <tag3>some text</tag3>
  </tag2>
</outer>
);
# nixExpr = { tag1 = "42"; tag2.tag3 = "some text"};
```

In the longer term the parser could gain schema validation and typing ("42" -> 42) but those are not currently required.

Are there already established way to achieve that without writing a nix implementation (xml2json -> fromJSON -> nix) ?
Is anyone working on something like that?

Thanks!

**Comment by vcunat** [comment]
Yes, I suppose the best way _for now_ is to use some generic xml2json with `builtins.fromJSON`.

**Comment by vcunat** [comment]
BTW, nix builtin issues actually belong into a [different repo](https://github.com/NixOS/nix), but here it's not 100% clear (to me) that the builtin approach will turn out to be the best one in this case as well.

**Comment by amazari** [comment]
@vcunat Thanks your answers.

> Yes, I suppose the best way _for now_ is to use some generic xml2json with `builtins.fromJSON`.

Excuse my lake of experience with nix, but is it possible to use an executable (ie. xml2json) from a derivation within some nix code? I mean, ideally I would like the list of available addons to be dynamic and not generated beforehand like
```
myKodi = kodi.withAddonsFrom "http://url-to-repo1" [
                       definitelyLegalContentAddon1,
                       starTrekLeetTheme
                    ]
                   .kodi.withAddonsFrom "http://url-to-repo2" [
                       ...
                       ...
                    ];
```
Here the list of available addons would be dynamically generated at runtime. Is something like that feasible ? Desirable? 

> BTW, nix builtin issues actually belong into a [different repo](https://github.com/NixOS/nix), but here it's not 100% clear (to me) whether the builtin approach will turn out to be the best one.

Indeed, this could be a pure nix implementation in the lib. Using builtin might make the implementation easier and sturdier by using a mature native parsing library, I guess.

Thanks!

**Comment by vcunat** [comment]
Importing from a derivation is... at least a non-standard thing ATM that we don't commonly allow in the NixPkgs repo.  I don't really know any details about it.  The main fundamental problem is that you need to "build"  (or at least download) already during evaluation, which is quite disruptive.

**Comment by vcunat** [comment]
Still, such conversion tools may be used e.g. in a generator for nix expressions that is _separately_ ran and the resulting `*.nix` (or json) gets committed into a repo.

**Comment by amazari** [comment]
Understood. I guess a recurring script preprocessing static package days
will do for now.
Plus it might provide more guarantees with versions compatibility by
pinning and avoid unexpected/unreproducible results.

Thanks for your guidance

Le dim. 10 mars 2019 à 14:49, Vladimír Čunát <notifications@github.com> a
écrit :

**Comment by amazari** [comment]
BTW how are run such tools (hackage2nix, emacs plugins generator..)?
Are they run periodically by maintainers or is there a hydra/bot/whatever automatic job?

**Comment by vcunat** [comment]
I think that varies.  For Haskell I think Peti runs the thing by hand.  @r-ryantm is fully automated, I suspect.

**Comment by milahu** [comment]
> achieve that without writing a nix implementation

sounds scary, but its easy with nix-parsec

xml parser in https://github.com/nprindle/nix-parsec/pull/6

more string parsers in https://nixos.wiki/wiki/String-parsing_in_Nix

> `nixExpr = { tag1 = "42"; tag2.tag3 = "some text"};`

every xml node is a mix of set and list (attributes and children)
that's why there is no `builtins.fromXML` and `builtins.toXML`

the mapping from XML to Nix looks more like

```yaml
- root
  children:
    - tag: outer
      children:
        - text: "\n  "
        - tag: tag1
          children:
            - text: "42"
        - text: "\n  "
        - tag: tag2
          children:
            - text: "\n    "
            - tag: tag3
              children:
                - text: "some text"
            - text: "\n  "
        - text: "\n"
```

(note to self: todo: write print-nix.nix)
